service/worker/replicator/historyRereplicator.go

```diff
@@ -0,0 +1,438 @@
+
+	runID := beginingRunID
+	for len(runID) != 0 && runID != endingRunID {
+		firstEventID, nextEventID := h.eventIDRange(runID, beginingRunID, beginingFirstEventID, endingRunID, endingNextEventID)
+		runID, err = h.sendSingleWorkflowHistory(domainID, workflowID, runID, firstEventID, nextEventID)
+		if err != nil {
+			return err
+		}
+	}
+
+	if runID == endingRunID {
+		firstEventID, nextEventID := h.eventIDRange(runID, beginingRunID, beginingFirstEventID, endingRunID, endingNextEventID)
+		_, err = h.sendSingleWorkflowHistory(domainID, workflowID, runID, firstEventID, nextEventID)
+		return err
+	}
+
+	// we have runID being empty string
+	// this means that beginingRunID to endingRunID does not have a continue as new relation ship
+	// such as beginingRunID -> runID1, then runID2 (no continued as new), then runID3 -> endingRunID
+	// need to work backwords using endingRunID,
+	// moreover, in the above case, runID2 cannot be replicated since no one points to it
+
+	// runIDs to be use to resend history, in reverse order
+	runIDs := []string{endingRunID}
+	runID = endingRunID
+	for len(runID) != 0 {
+		runID, err = h.getPrevRunID(domainID, workflowID, runID)
+		if err != nil {
+			return err
+		}
+		runIDs = append(runIDs, runID)
+	}
+
+	// the last runID append in the array is empty
+	// for all the runIDs, send the history
+	for index := len(runIDs) - 2; index > -1; index-- {
+		runID = runIDs[index]
+		firstEventID, nextEventID := h.eventIDRange(runID, beginingRunID, beginingFirstEventID, endingRunID, endingNextEventID)
+		_, err = h.sendSingleWorkflowHistory(domainID, workflowID, runID, firstEventID, nextEventID)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (h *HistoryRereplicatorImpl) sendSingleWorkflowHistory(domainID string, workflowID string, runID string,
+	firstEventID int64, nextEventID int64) (string, error) {
+
+	if firstEventID == nextEventID {
+		// this is handling the case where the first event of a workflow triggers a resend
+		return "", nil
+	}
+
+	var request *history.ReplicateRawEventsRequest // pending replication request to history, initialized to nil
+
+	// branch token, event store version, replication
+	// for each request to history
+	var branchToken []byte
+	var eventStoreVersion int32
+	var replicationInfo map[string]*internal.ReplicationInfo
+
+	pageSize := int32(100)
+	var token []byte
+	for doPaging := true; doPaging; doPaging = len(token) > 0 {
+		response, err := h.getHistory(domainID, workflowID, runID, branchToken, firstEventID, nextEventID, token, pageSize)
+		if err != nil {
+			return "", err
+		}
+
+		branchToken = response.BranchToken
+		eventStoreVersion = response.GetEventStoreVersion()
+		replicationInfo = h.replicationInfoFromPublic(response.ReplicationInfo)
+		token = response.NextPageToken
+
+		for _, externalBatch := range response.HistoryBatches {
+			err := h.sendReplicationRawRequest(request)
```

Comment:
the first request is nil, yes it is handled in the function, but it is very confusing to read this code. is it possible to re-organize the code so it prepare the request first and then send it after, or at least adding some more comments.

Reply:
comment added